### PR TITLE
Enum validator improvements

### PR DESCRIPTION
**Pull request by samuelcolvin**
## Change Summary

While looking through how we do enum validation in preparation for moving some of it to Rust, I saw a few things that could be improved. Hopefully this provides a "cleaner" schema, and should improve performance a bit.

Performance gain is small, but not insignificant:

### before

```py
In [1]: import enum
   ...: from pydantic import TypeAdapter
   ...: 
   ...: class Foo(enum.IntEnum):
   ...:     x = 1
   ...:     y = 2
   ...: 
   ...: 
   ...: ta = TypeAdapter(Foo)
   ...: 
   ...: %timeit ta.validate_python(2)
   ...: %timeit ta.validate_json('2')
   ...: 
550 ns ± 2.37 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
560 ns ± 2.29 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

### After

```py
In [1]: import enum
   ...: from pydantic import TypeAdapter
   ...: 
   ...: class Foo(enum.IntEnum):
   ...:     x = 1
   ...:     y = 2
   ...: 
   ...: 
   ...: ta = TypeAdapter(Foo)
   ...: 
   ...: %timeit ta.validate_python(2)
   ...: %timeit ta.validate_json('2')
   ...: 
508 ns ± 0.84 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
517 ns ± 1.95 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: pydantic/_internal/_std_types_schema.py [hunk 1]
```diff
@@ -90,47 +91,32 @@ def to_enum(input_value: Any, /) -> Enum:
         try:
             return enum_type(input_value)
         except ValueError:
-            # The type: ignore on the next line is to ignore the requirement of LiteralString
-            raise PydanticCustomError('enum', f'Input should be {expected}', {'expected': expected})  # type: ignore
+            raise PydanticCustomError('enum', 'Input should be {expected}', {'expected': expected})
```
Comment (samuelcolvin) [inline pydantic/_internal/_std_types_schema.py:94]: output is the same, but it's more correct to let pydantic-core take care of replacing the variables in the message template.

File: pydantic/_internal/_std_types_schema.py [hunk 2]
```diff
@@ -90,47 +91,32 @@ def to_enum(input_value: Any, /) -> Enum:
...
         # this handles `IntEnum`, and also `Foobar(int, Enum)`
-        updates['type'] = 'integer'
-        lax = core_schema.chain_schema([core_schema.int_schema(), to_enum_validator])
-        # Disallow float from JSON due to strict mode
-        strict = core_schema.json_or_python_schema(
-            json_schema=core_schema.no_info_after_validator_function(to_enum, core_schema.int_schema()),
-            python_schema=strict_python_schema,
-        )
+        js_updates['type'] = 'integer'
+        lax_schema = core_schema.no_info_after_validator_function(to_enum, core_schema.int_schema())
```
Comment (samuelcolvin) [inline pydantic/_internal/_std_types_schema.py:99]: I've replaced lots of use of `core_schema.chain_schema` with `core_schema.no_info_after_validator_function` which should be more efficient, and is more concise.

File: pydantic/_internal/_std_types_schema.py [hunk 3]
```diff
@@ -90,47 +91,32 @@ def to_enum(input_value: Any, /) -> Enum:
...
+        ref=enum_ref,
+        metadata={'pydantic_js_functions': [get_json_schema]},
     )
-    use_enum_values = config.get('use_enum_values', False)
-    if use_enum_values:
-        enum_schema = core_schema.chain_schema(
-            [enum_schema, core_schema.no_info_plain_validator_function(lambda x: x.value)]
-        )
+    if config.get('use_enum_values', False):
+        enum_schema = core_schema.no_info_after_validator_function(attrgetter('value'), enum_schema)
```
Comment (samuelcolvin) [inline pydantic/_internal/_std_types_schema.py:119]: `attrgetter` is nearly twice as fast as a lambda:

```python
In [1]: import enum

In [2]: class Foo(enum.Enum):
   ...:     x = 'a'
   ...:     y = 'b'
   ...: 

In [3]: with_lambda = lambda thing: thing.x

In [4]: %timeit with_lambda(Foo)
39.1 ns ± 0.161 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [5]: from operator import attrgetter

In [6]: using_attrgetter = attrgetter('x')

In [7]: %timeit using_attrgetter(Foo)
24.5 ns ± 0.473 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

Comment (sydney-runkle) [reply pydantic/_internal/_std_types_schema.py:119]: Should we be using `attrgetter` more broadly?

**Review by sydney-runkle — APPROVED** [review]
Looks great! This is much more simple than the existing behavior!

**Pull request merged**